### ISSUE TITLE
Version fragmentation problem

### ISSUE THREAD

**Issue by Noviny** [atlassian/changesets#189]
Wanted to create a clear place to discuss this. Going to try and outline what the problem *is* in this description, then going to come back in comments to outline different ways of approaching it, and my idea of how changesets should or shouldn't address this problem.

> This problem exists in any instance where you release multiple interdependent packages, and scales with the interdependency of the packages.

## The problem

In the simplest form, the problem is that in an interdependent system, you may end up pulling in multiple copies of the same package.

Our first hypothetical:

I am publishing a button component and a modal component. The modal depends on button.

In essence, I get:

```
button@1.0.0
modal@1.00
    button@^1.0.0
```

If a user installs both of these packages, modern package managers will 'dedupe' the versions of button, meaning the user has one version.

If however we have:

```
button@1.1.0
modal@1.00
    button@^1.0.0
```

If the above user upgrades button, but leaves modal alone, modern package managers will keep the button modal installed 'locked' at `1.0.0` and pull in a new version of button for `button` itself.

The user now has two copies of the button code that may be nearly identical, and that the semver-ing of modal's dependency on button suggests that there should be no problem with modal using the updated button.

There are two ways this spread occurs:

- installing packages released at different times so dependencies cannot be resolved to a single version
- installing packages that get 'locked' to resolving transitive dependencies to different versions despite them being semver compatible to dedupe.

## How big of a problem is it?

This problem scales very badly with system interconnectivity, system size and release cadence.

Basically the more versions of more packages you have, which have more interdependencies, the worse it will get.

Some monorepo-based projects won't even notice this, or will be completely unaffected. Others may find it a massive drain on their ability to be installed and maintained.

## Notes

The problem you are trying to solve for is an interrelation between consumer install habits and publisher publish habits.

Most solutions will have trade-offs.

**Comment by Noviny** [comment]
(I need to head home so I'm going to come back and fill this out in a bit)

## Ways to lessen the problem

### Fewer packages (less interdependency through combining single packages together)

### Fewer versions of each package (more changes per version)

###

**Comment by Noviny** [comment]
(I am considering making this a PR just so I can start making threads of comments off an md doc)

**Comment by Noviny** [comment]
@lukebatchelor you mentioned wanting a larger talk about version fragmentation so I made a thing.

@mitchellhamilton you probably care.

**Comment by bluwy** [comment]
I believe https://changesets.dev/guide/config#updateinternaldependencies handles this problem today.
